go/workflow-interceptor-updates.md

```diff
@@ -0,0 +1,269 @@
+
+### WorkflowXCallsInterceptor
+
+Proposal: Change the name to `WorkflowXInterceptor`. Why: `Calls` just adds noise (more accepted in Java, not Go).
+
+### WorkflowInboundCallsInterceptor.ExecuteWorkflow
+
+Current:
+
+```go
+ExecuteWorkflow(ctx Context, workflowType string, args ...interface{}) []interface{}
+```
+
+Proposal:
+
+```go
+type WorkflowInput struct {
+  Args []interface{}
+}
+
+ExecuteWorkflow(ctx Context, input *WorkflowInput) (interface{}, error)
+```
+
+Why:
+
+* Workflow type can be derived from the context _or_ the info first given to `InterceptWorkflow`
+* Workflows can only return a value + error or just an error and this makes it more explicit what the error is
+
+### WorkflowInboundCallsInterceptor.ProcessSignal
+
+Current:
+
+```go
+ProcessSignal(ctx Context, signalName string, arg interface{}) error
+```
+
+Proposal:
+
+```go
+type SignalInput struct {
+  SignalName string
+  Arg interface{}
+  EventID string
+}
+
+HandleSignal(ctx Context, input *SignalInput) error
+```
+
+Also, this will be called to handle the signal with the last invocation in the chain passing the signal to the workflow.
+
+Why:
+
+* `HandleSignal` matches Java nomenclature and new behavior better
+* Currently sending is sent separate and async, so the current version doesn't wrap the handling. Instead this will be
+  part of the async call and documentation will make clear it needs to be safe for concurrent invocation.
+* The Java version provides the event ID so we will too
+
+### WorfkflowInboundCallsInterceptor.HandleQuery
+
+Current:
+
+```go
+HandleQuery(ctx Context, queryType string, args *commonpb.Payloads,
+  handler func(*commonpb.Payloads) (*commonpb.Payloads, error)) (*commonpb.Payloads, error)
+```
+
+Proposal:
+
+```go
+type QueryInput struct {
+  QueryType string
+  Args []interface{}
+}
+
+HandleQuery(ctx Context, input *QueryInput) (interface{}, error)
+```
+
+Why:
+
+* All other calls convert the payload outside of the interceptor call, this can too
+* No need to pass a handler, that is just part of the last item in the interceptor chain (i.e. when calling next)
+
+### WorkflowOutboundCallsInterceptor
+
+We can leave everything the way it is because, while it doesn't make the interceptor parameters future-proof with struct
+wrappers, it mimics the actual calls that aren't going to have backwards-compatible alterations.
+
+### WorkflowOutboundCallsInterceptor.GetWorkflowInfo
+
+Current:
+
+```go
+GetWorkflowInfo(ctx Context) *WorkflowInfo
+```
+
+Proposal:
+
+```go
+GetInfo(ctx Context) *WorkflowInfo
+```
+
+Why: All other calls match `workflow` package, why not this?
+
+### ActivityInterceptor
+
+Proposal:
+
+```go
+type ActivityInterceptor interface {
+  InterceptActivity(ctx context.Context, next ActivityInboundInterceptor) ActivityInboundInterceptor
+}
+```
+
+Why: Mimics `WorkflowInterceptor` but for activities.
+
+### ActivityInboundInterceptor
+
+Proposal:
+
+```go
+type ActivityInput struct {
+  Args []interface{}
```

Comment:
I'd put the activity type and headers here so they can be modified by the interceptor.

Reply:
Headers are modified in the context usually and I will show examples of that. As for activity type, that's not a feature I see of the Java SDK. Is this something we need to support in interceptors in both places? My goal was to reach feature parity with Java but I can add more if wanted.

Reply:
I just saw that the type in missing in Java and Node, I'm wondering if we should add it or not and if it makes sense for that to be mutable by interceptors.

Reply:
This struct supports additional items in the future, I don't think such a decision must be made now.